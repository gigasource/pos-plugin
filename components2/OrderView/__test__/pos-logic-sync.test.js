//<editor-fold desc="declare">
import {
  addPayment,
  cancelOrder,
  changeCourse,
  clearPayment,
  removeItem,
  removeModifier,
  updateItem,
  hooks,
  simulateBackendPrint,
  changeItemQuantity
} from "../pos-logic";
import {nextTick} from "vue";

import {
  addProduct,
  orderBeFactory
} from '../pos-logic-be'
import {mockProducts} from "./mock_products";
import _ from 'lodash';

const orderMaster = orderBeFactory(2)
const orderClient = orderBeFactory(3)

const {syncFactory} = require('../../../test-utils')
const {stringify} = require('schemahandler/utils')

jest.setTimeout(60000)

//<editor-fold desc="prepare test">
const {
  prepareActionCommitTest
} = require('../../../backend/commit/actionCommit.prepare.test')
const {
  prepareOrderTest,
  checkOrderCreated
} = require('../../../backend/order/order.prepare.test')
const {
  prepareOrderCommit
} = require('../../../backend/commit/orderCommit.prepare.test')
require("mockdate").set(new Date("2021-01-01").getTime());
//</editor-fold>

let cms
//</editor-fold>

/**
 * cms[0] always is master
 */
beforeAll(async () => {
  cms = await syncFactory('posLogicSync', 2)

  for (let _cms of cms) {
    prepareActionCommitTest(_cms)
    prepareOrderCommit(_cms)
    prepareOrderTest(_cms)
    _cms.triggerFeConnect()
  }
})

beforeEach(async () => {
  for (let _cms of cms) {
    await _cms.orm('Order').remove({}).direct()
    await _cms.orm('Commit').remove({}).direct()
    await _cms.orm('Action').remove({}).direct()
  }
})

let [cola, pepsi, water, soda] = mockProducts;
const convert = actionList => {
  return actionList.value.filter(i => i.action[0].fn === 'findOneAndUpdate' && i.action[0].args[1].$push).map(i => i.action[0].args[1].$push.items)
}

/**
 * cms[0] ~ orderMaster
 * cms[1] ~ orderClient
 */
describe('Pos logic sync', function () {
  it('Case 1: Create order + addProduct in master', async (done) => {
    orderMaster.prepareOrder('10')
    const order1 = orderMaster.getCurrentOrder()
    addProduct(order1, water);
    simulateBackendPrint(order1);
    orderClient.prepareOrder(_.cloneDeep(order1));
    const order2 = orderClient.getCurrentOrder()
    cms[1].feSocket.on('update-table', async function (order) {
      await checkOrderCreated(cms[0].orm)
      await checkOrderCreated(cms[1].orm)
      await orderClient.syncOrderChange(order)
      //todo: fix condition :uuid
      expect(stringify(orderMaster.getCurrentOrder())).toMatchSnapshot()
      expect(stringify(orderClient.getCurrentOrder())).toMatchSnapshot()
      expect(stringify(orderClient.actionList.value)).toMatchSnapshot()
      done()
    })
    await nextTick()
    addProduct(order1, cola)
    await nextTick()
    addProduct(order1, soda)
    await nextTick()
    changeItemQuantity(order2, 0, 2);
    addProduct(order2, pepsi)
    await nextTick()
    expect(stringify(convert(orderMaster.actionList))).toMatchSnapshot()
    expect(stringify(orderClient.actionList.value)).toMatchSnapshot()
    await nextTick()
    simulateBackendPrint(order1)
    cms[0].feSocket.emit('print-to-kitchen', orderMaster.actionList.value, order1);
    await nextTick()
  })
})
