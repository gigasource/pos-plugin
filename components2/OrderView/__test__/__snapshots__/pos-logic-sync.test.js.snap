// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pos logic sync Case 1: Create order + addProduct in master 1`] = `
Array [
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
            "cancellationItems": Array [],
            "items": Array [],
            "payment": Array [],
            "status": "inProgress",
            "table": "10",
            "takeAway": false,
            "user": Array [],
            "vDate": "2021-01-20T17:00:00.000Z",
            "vSum": 0,
            "vTaxSum": Object {},
          },
        ],
        "fn": "create",
      },
      Object {
        "args": Array [
          "createOrder",
          Object {
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$push": Object {
              "items": Object {
                "_id": "ObjectID",
                "course": 1,
                "groupPrinter": "",
                "id": "3",
                "name": "Water",
                "price": 67,
                "printed": true,
                "product": "ObjectID",
                "quantity": 1,
                "sent": true,
                "tax": 7,
                "taxes": Array [
                  7,
                  null,
                ],
                "vSum": 67,
                "vTakeAway": false,
                "vTaxSum": Object {
                  "7": Object {
                    "gross": 67,
                    "net": 62.62,
                    "tax": 4.38,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$set": Object {
              "vSum": 67,
              "vTaxSum": Object {
                "7": Object {
                  "gross": 67,
                  "net": 62.62,
                  "tax": 4.38,
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$push": Object {
              "items": Object {
                "_id": "ObjectID",
                "course": 1,
                "groupPrinter": "",
                "id": 1,
                "name": "Cola",
                "price": 98,
                "quantity": 1,
                "tax": 19,
                "taxes": Array [
                  19,
                  null,
                ],
                "vSum": 98,
                "vTakeAway": false,
                "vTaxSum": Object {
                  "19": Object {
                    "gross": 98,
                    "net": 82.35,
                    "tax": 15.65,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$set": Object {
              "vSum": 165,
              "vTaxSum": Object {
                "19": Object {
                  "gross": 98,
                  "net": 82.35,
                  "tax": 15.65,
                },
                "7": Object {
                  "gross": 67,
                  "net": 62.62,
                  "tax": 4.38,
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 2`] = `
Array [
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
            "items._id": "ObjectID",
          },
          Object {
            "$set": Object {
              "items.$": Object {
                "vSum": 201,
                "vTaxSum": Object {
                  "7": Object {
                    "gross": 201,
                    "net": 187.85,
                    "tax": 13.15,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$push": Object {
              "items": Object {
                "_id": "ObjectID",
                "course": 1,
                "groupPrinter": "",
                "id": 2,
                "name": "Pepsi",
                "price": 35,
                "quantity": 1,
                "tax": 0,
                "taxes": Array [
                  0,
                  null,
                ],
                "vSum": 35,
                "vTakeAway": false,
                "vTaxSum": Object {
                  "0": Object {
                    "gross": 35,
                    "net": 35,
                    "tax": 0,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$set": Object {
              "vSum": 236,
              "vTaxSum": Object {
                "0": Object {
                  "gross": 35,
                  "net": 35,
                  "tax": 0,
                },
                "7": Object {
                  "gross": 201,
                  "net": 187.85,
                  "tax": 13.15,
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$push": Object {
              "items": Object {
                "_id": "ObjectID",
                "course": 1,
                "groupPrinter": "",
                "id": "3",
                "name": "Water",
                "originalQuantity": 1,
                "price": 67,
                "printed": true,
                "product": "ObjectID",
                "quantity": 3,
                "sent": true,
                "tax": 7,
                "taxes": Array [
                  7,
                  null,
                ],
                "vSum": 201,
                "vTakeAway": false,
                "vTaxSum": Object {
                  "7": Object {
                    "gross": 201,
                    "net": 187.85,
                    "tax": 13.15,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$push": Object {
              "items": Object {
                "_id": "ObjectID",
                "course": 1,
                "groupPrinter": "",
                "id": 2,
                "name": "Pepsi",
                "price": 35,
                "quantity": 1,
                "tax": 0,
                "taxes": Array [
                  0,
                  null,
                ],
                "vSum": 35,
                "vTakeAway": false,
                "vTaxSum": Object {
                  "0": Object {
                    "gross": 35,
                    "net": 35,
                    "tax": 0,
                  },
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
  Object {
    "action": Array [
      Object {
        "args": Array [
          Object {
            "_id": "ObjectID",
          },
          Object {
            "$set": Object {
              "_id": "ObjectID",
              "payment": Array [],
              "status": "inProgress",
              "table": "10",
              "takeAway": false,
              "user": Array [],
              "vDate": "2021-01-20T17:00:00.000Z",
              "vSum": 236,
              "vTaxSum": Object {
                "0": Object {
                  "gross": 35,
                  "net": 35,
                  "tax": 0,
                },
                "7": Object {
                  "gross": 201,
                  "net": 187.85,
                  "tax": 13.15,
                },
              },
            },
          },
        ],
        "fn": "findOneAndUpdate",
      },
      Object {
        "args": Array [
          "updateActiveOrder",
          Object {
            "orderId": "ObjectID",
            "table": "10",
          },
        ],
        "fn": "commit",
      },
    ],
    "modelName": "Order",
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 3`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": "3",
        "name": "Water",
        "price": 67,
        "printed": true,
        "product": "ObjectID",
        "quantity": 1,
        "sent": true,
        "tax": 7,
        "taxes": Array [
          7,
          null,
        ],
        "vSum": 67,
        "vTakeAway": false,
        "vTaxSum": Object {
          "7": Object {
            "gross": 67,
            "net": 62.62,
            "tax": 4.38,
          },
        },
      },
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": 1,
        "name": "Cola",
        "price": 98,
        "printed": true,
        "quantity": 1,
        "sent": true,
        "tax": 19,
        "taxes": Array [
          19,
          null,
        ],
        "vSum": 98,
        "vTakeAway": false,
        "vTaxSum": Object {
          "19": Object {
            "gross": 98,
            "net": 82.35,
            "tax": 15.65,
          },
        },
      },
    ],
    "payment": Array [],
    "status": "inProgress",
    "table": "10",
    "takeAway": false,
    "user": Array [],
    "vDate": "2021-01-20T17:00:00.000Z",
    "vSum": 165,
    "vTaxSum": Object {
      "19": Object {
        "gross": 98,
        "net": 82.35,
        "tax": 15.65,
      },
      "7": Object {
        "gross": 67,
        "net": 62.62,
        "tax": 4.38,
      },
    },
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 4`] = `
Array [
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"insertOne\\",\\"args\\":[{\\"items\\":[],\\"cancellationItems\\":[],\\"payment\\":[],\\"user\\":[],\\"table\\":\\"10\\",\\"_id\\":\\"ObjectID\\",\\"takeAway\\":false,\\"status\\":\\"inProgress\\",\\"vSum\\":0,\\"vDate\\":\\"2021-01-20T17:00:00.000Z\\",\\"vTaxSum\\":{}}]}]",
    "collectionName": "Order",
    "condition": "null",
    "data": Object {
      "docId": "ObjectID",
      "table": "10",
    },
    "fromMaster": true,
    "id": 1,
    "tags": Array [
      "createOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$push\\":{\\"items\\":{\\"groupPrinter\\":\\"\\",\\"id\\":\\"3\\",\\"name\\":\\"Water\\",\\"price\\":67,\\"product\\":\\"ObjectID\\",\\"taxes\\":[7,null],\\"course\\":1,\\"quantity\\":1,\\"_id\\":\\"ObjectID\\",\\"sent\\":true,\\"printed\\":true,\\"vSum\\":67,\\"vTakeAway\\":false,\\"tax\\":7,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67}}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "fromMaster": true,
    "id": 2,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"vSum\\":67,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "fromMaster": true,
    "id": 3,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$push\\":{\\"items\\":{\\"groupPrinter\\":\\"\\",\\"id\\":1,\\"name\\":\\"Cola\\",\\"price\\":98,\\"taxes\\":[19,null],\\"course\\":1,\\"quantity\\":1,\\"_id\\":\\"ObjectID\\",\\"vSum\\":98,\\"vTakeAway\\":false,\\"tax\\":19,\\"vTaxSum\\":{\\"19\\":{\\"tax\\":15.65,\\"net\\":82.35,\\"gross\\":98}}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "fromMaster": true,
    "id": 4,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"vSum\\":165,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67},\\"19\\":{\\"tax\\":15.65,\\"net\\":82.35,\\"gross\\":98}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "fromMaster": true,
    "id": 5,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"items.$[].sent\\":true,\\"items.$[].printed\\":true,\\"cancellationItems.$[].sent\\":true,\\"cancellationItems.$[].printed\\":true}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
    },
    "fromMaster": true,
    "id": 6,
    "tags": Array [],
    "uuid": "uuid-v1",
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 5`] = `
Array [
  Object {
    "_id": "ObjectID",
    "cancellationItems": Array [],
    "items": Array [
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": "3",
        "name": "Water",
        "price": 67,
        "printed": true,
        "product": "ObjectID",
        "quantity": 1,
        "sent": true,
        "tax": 7,
        "taxes": Array [
          7,
          null,
        ],
        "vSum": 67,
        "vTakeAway": false,
        "vTaxSum": Object {
          "7": Object {
            "gross": 67,
            "net": 62.62,
            "tax": 4.38,
          },
        },
      },
      Object {
        "_id": "ObjectID",
        "course": 1,
        "groupPrinter": "",
        "id": 1,
        "name": "Cola",
        "price": 98,
        "printed": true,
        "quantity": 1,
        "sent": true,
        "tax": 19,
        "taxes": Array [
          19,
          null,
        ],
        "vSum": 98,
        "vTakeAway": false,
        "vTaxSum": Object {
          "19": Object {
            "gross": 98,
            "net": 82.35,
            "tax": 15.65,
          },
        },
      },
    ],
    "payment": Array [],
    "status": "inProgress",
    "table": "10",
    "takeAway": false,
    "user": Array [],
    "vDate": "2021-01-20T17:00:00.000Z",
    "vSum": 165,
    "vTaxSum": Object {
      "19": Object {
        "gross": 98,
        "net": 82.35,
        "tax": 15.65,
      },
      "7": Object {
        "gross": 67,
        "net": 62.62,
        "tax": 4.38,
      },
    },
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 6`] = `
Array [
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"insertOne\\",\\"args\\":[{\\"items\\":[],\\"cancellationItems\\":[],\\"payment\\":[],\\"user\\":[],\\"table\\":\\"10\\",\\"_id\\":\\"ObjectID\\",\\"takeAway\\":false,\\"status\\":\\"inProgress\\",\\"vSum\\":0,\\"vDate\\":\\"2021-01-20T17:00:00.000Z\\",\\"vTaxSum\\":{}}]}]",
    "collectionName": "Order",
    "condition": "null",
    "data": Object {
      "docId": "ObjectID",
      "table": "10",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 1,
    "tags": Array [
      "createOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$push\\":{\\"items\\":{\\"groupPrinter\\":\\"\\",\\"id\\":\\"3\\",\\"name\\":\\"Water\\",\\"price\\":67,\\"product\\":\\"ObjectID\\",\\"taxes\\":[7,null],\\"course\\":1,\\"quantity\\":1,\\"_id\\":\\"ObjectID\\",\\"sent\\":true,\\"printed\\":true,\\"vSum\\":67,\\"vTakeAway\\":false,\\"tax\\":7,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67}}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 2,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"vSum\\":67,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 3,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$push\\":{\\"items\\":{\\"groupPrinter\\":\\"\\",\\"id\\":1,\\"name\\":\\"Cola\\",\\"price\\":98,\\"taxes\\":[19,null],\\"course\\":1,\\"quantity\\":1,\\"_id\\":\\"ObjectID\\",\\"vSum\\":98,\\"vTakeAway\\":false,\\"tax\\":19,\\"vTaxSum\\":{\\"19\\":{\\"tax\\":15.65,\\"net\\":82.35,\\"gross\\":98}}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 4,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"vSum\\":165,\\"vTaxSum\\":{\\"7\\":{\\"tax\\":4.38,\\"net\\":62.62,\\"gross\\":67},\\"19\\":{\\"tax\\":15.65,\\"net\\":82.35,\\"gross\\":98}}}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
      "orderId": "ObjectID",
      "table": "10",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 5,
    "tags": Array [
      "updateActiveOrder",
    ],
    "uuid": "uuid-v1",
  },
  Object {
    "_id": "ObjectID",
    "approved": false,
    "chain": "[{\\"fn\\":\\"findOneAndUpdate\\",\\"args\\":[{\\"_id\\":\\"ObjectID\\"},{\\"$set\\":{\\"items.$[].sent\\":true,\\"items.$[].printed\\":true,\\"cancellationItems.$[].sent\\":true,\\"cancellationItems.$[].printed\\":true}}]},{\\"fn\\":\\"setOptions\\",\\"args\\":[{\\"new\\":true}]}]",
    "collectionName": "Order",
    "condition": "{\\"_id\\":\\"ObjectID\\"}",
    "data": Object {
      "docId": "ObjectID",
    },
    "dbName": null,
    "fromMaster": true,
    "id": 6,
    "tags": Array [],
    "uuid": "uuid-v1",
  },
]
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 7`] = `
Object {
  "_id": "ObjectID",
  "cancellationItems": Array [],
  "items": Array [
    Object {
      "_id": "ObjectID",
      "course": 1,
      "groupPrinter": "",
      "id": "3",
      "name": "Water",
      "price": 67,
      "printed": true,
      "product": "ObjectID",
      "quantity": 1,
      "sent": true,
      "tax": 7,
      "taxes": Array [
        7,
        null,
      ],
      "vSum": 67,
      "vTakeAway": false,
      "vTaxSum": Object {
        "7": Object {
          "gross": 67,
          "net": 62.62,
          "tax": 4.38,
        },
      },
    },
    Object {
      "_id": "ObjectID",
      "course": 1,
      "groupPrinter": "",
      "id": 1,
      "name": "Cola",
      "price": 98,
      "printed": true,
      "quantity": 1,
      "sent": true,
      "tax": 19,
      "taxes": Array [
        19,
        null,
      ],
      "vSum": 98,
      "vTakeAway": false,
      "vTaxSum": Object {
        "19": Object {
          "gross": 98,
          "net": 82.35,
          "tax": 15.65,
        },
      },
    },
  ],
  "payment": Array [],
  "status": "inProgress",
  "table": "10",
  "takeAway": false,
  "user": Array [],
  "vDate": "2021-01-20T17:00:00.000Z",
  "vSum": 165,
  "vTaxSum": Object {
    "19": Object {
      "gross": 98,
      "net": 82.35,
      "tax": 15.65,
    },
    "7": Object {
      "gross": 67,
      "net": 62.62,
      "tax": 4.38,
    },
  },
}
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 8`] = `
Object {
  "_id": "ObjectID",
  "cancellationItems": Array [],
  "items": Array [
    Object {
      "_id": "ObjectID",
      "course": 1,
      "groupPrinter": "",
      "id": "3",
      "name": "Water",
      "originalQuantity": 1,
      "price": 67,
      "printed": true,
      "product": "ObjectID",
      "quantity": 3,
      "sent": true,
      "tax": 7,
      "taxes": Array [
        7,
        null,
      ],
      "vSum": 201,
      "vTakeAway": false,
      "vTaxSum": Object {
        "7": Object {
          "gross": 201,
          "net": 187.85,
          "tax": 13.15,
        },
      },
    },
    Object {
      "_id": "ObjectID",
      "course": 1,
      "groupPrinter": "",
      "id": 1,
      "name": "Cola",
      "originalQuantity": 1,
      "price": 98,
      "printed": true,
      "quantity": 1,
      "sent": true,
      "tax": 19,
      "taxes": Array [
        19,
        null,
      ],
      "vSum": 98,
      "vTakeAway": false,
      "vTaxSum": Object {
        "19": Object {
          "gross": 98,
          "net": 82.35,
          "tax": 15.65,
        },
      },
    },
    Object {
      "_id": "ObjectID",
      "course": 1,
      "groupPrinter": "",
      "id": 2,
      "name": "Pepsi",
      "price": 35,
      "quantity": 1,
      "tax": 0,
      "taxes": Array [
        0,
        null,
      ],
      "vSum": 35,
      "vTakeAway": false,
      "vTaxSum": Object {
        "0": Object {
          "gross": 35,
          "net": 35,
          "tax": 0,
        },
      },
    },
  ],
  "payment": Array [],
  "status": "inProgress",
  "table": "10",
  "takeAway": false,
  "user": Array [],
  "vDate": "2021-01-20T17:00:00.000Z",
  "vSum": 334,
  "vTaxSum": Object {
    "0": Object {
      "gross": 35,
      "net": 35,
      "tax": 0,
    },
    "19": Object {
      "gross": 98,
      "net": 82.35,
      "tax": 15.65,
    },
    "7": Object {
      "gross": 201,
      "net": 187.85,
      "tax": 13.15,
    },
  },
}
`;

exports[`Pos logic sync Case 1: Create order + addProduct in master 9`] = `Array []`;
